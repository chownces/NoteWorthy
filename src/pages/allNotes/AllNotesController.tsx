import { gql, useQuery } from '@apollo/client';
import React from 'react';

import AllNotes, { AllNotesProps } from './AllNotes';

// TODO: Clean up the various type exports...
// `blocks` is excluded here as we do not need it when displaying AllNotes for now.
export type Note = {
  _id: string;
  title: string;
  date: string;
};

const AllNotesController: React.FC = () => {
  // TODO: Handle repositioning of Note 'blocks' in AllNotes page

  const [notes, setNotes] = React.useState<Note[]>([]);

  // TODO: Change date to reflect the date and time of the latest changes (requires changes in backend)
  const GET_ALL_NOTES_QUERY = gql`
    {
      allNotes {
        _id
        title
        date
      }
    }
  `;

  const { loading, error, data } = useQuery(GET_ALL_NOTES_QUERY);

  /**
   * This effect handles the rerendering of the page once data is fetched asynchronously.
   */
  React.useEffect(() => {
    if (!loading) {
      if (!error) {
        setNotes(data.allNotes);
      } else {
        // TODO: Handle request error (e.g. display message to user)
      }
    }
  }, [loading, error, data, setNotes]);

  const allNotesProps: AllNotesProps = {
    notes: notes
  };

  return <AllNotes {...allNotesProps} />;
};

export default AllNotesController;
