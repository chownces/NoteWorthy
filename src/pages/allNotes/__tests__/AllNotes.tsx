import { MockedProvider } from '@apollo/client/testing';
import { act, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import AllNotesController, {
  CREATE_NOTE_MUTATION,
  GET_ALL_NOTES_QUERY
} from '../AllNotesController';

test('Gets all notes successfully and renders them correctly', async () => {
  // TODO: Relook this once authentication and filters are added
  const mocks = [
    {
      request: {
        query: GET_ALL_NOTES_QUERY
        // variables: {
        //   name: 'Buck',
        // },
      },
      result: {
        data: {
          allNotes: [
            {
              id: '6098ba47fdaba558e1955a00',
              title: 'test note 1',
              date: '21/05/21'
              // blocks: [
              //   {
              //     id: "koi4onszhkbdbzj9hm",
              //     html: "block1",
              //     tag: "p"
              //   },
              //   {
              //     id: "koi4fkuoqkqkg978gun",
              //     html: "block2",
              //     tag: "p"
              //   },
              //   {
              //     id: "koi8fn3o3od7ajdlw6v",
              //     html: "block3",
              //     tag: "p"
              //   },
              //   {
              //     id: "koic9efk29q64v8vowg",
              //     html: "",
              //     tag: "p"
              //   }
              // ]
            },
            {
              id: '6098bbdc04c0fa63070bfe5e',
              title: 'untitled',
              date: '31/12/21'
              // blocks: [
              //   {
              //     id: "koi4o9e370nwp2offt",
              //     html: " ",
              //     tag: "p"
              //   }
              // ]
            }
          ]
        }
      }
    }
  ];

  const { container: tree } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <AllNotesController />
      </MemoryRouter>
    </MockedProvider>
  );

  await act(() => new Promise(resolve => setTimeout(resolve, 0)));

  expect(tree).toMatchSnapshot();
});

test('Renders loading state correctly', async () => {
  // TODO: Relook this once authentication and filters are added
  const mocks = [
    {
      request: {
        query: GET_ALL_NOTES_QUERY
        // variables: {
        //   name: 'Buck',
        // },
      },
      result: {
        data: {
          allNotes: [
            {
              id: '6098ba47fdaba558e1955a00',
              title: 'test note 1',
              date: '21/05/21'
              // blocks: [
              //   {
              //     id: "koi4onszhkbdbzj9hm",
              //     html: "block1",
              //     tag: "p"
              //   },
              //   {
              //     id: "koi4fkuoqkqkg978gun",
              //     html: "block2",
              //     tag: "p"
              //   },
              //   {
              //     id: "koi8fn3o3od7ajdlw6v",
              //     html: "block3",
              //     tag: "p"
              //   },
              //   {
              //     id: "koic9efk29q64v8vowg",
              //     html: "",
              //     tag: "p"
              //   }
              // ]
            },
            {
              id: '6098bbdc04c0fa63070bfe5e',
              title: 'untitled',
              date: '31/12/21'
              // blocks: [
              //   {
              //     id: "koi4o9e370nwp2offt",
              //     html: " ",
              //     tag: "p"
              //   }
              // ]
            }
          ]
        }
      }
    }
  ];

  const { container: tree } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <AllNotesController />
      </MemoryRouter>
    </MockedProvider>
  );

  expect(tree).toMatchSnapshot();
});

test('Handles network error correctly ', async () => {
  // TODO: Relook this once authentication and filters are added
  const mocks = [
    {
      request: {
        query: GET_ALL_NOTES_QUERY
        // variables: {
        //   name: 'Buck',
        // },
      },
      error: new Error('An error occurred')
    }
  ];

  const { container: tree } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <AllNotesController />
      </MemoryRouter>
    </MockedProvider>
  );

  await act(() => new Promise(resolve => setTimeout(resolve, 0)));

  expect(tree).toMatchSnapshot();
});

// TODO: Add tests for update note mutation
