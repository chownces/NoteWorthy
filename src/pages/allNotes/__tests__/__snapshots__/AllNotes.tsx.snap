// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gets all notes successfully and renders them correctly 1`] = `
<div>
  <button>
    New Note
  </button>
  <div>
    <a
      href="/note/6098ba47fdaba558e1955a00"
    >
      <div
        class="allnotes-note"
      >
        <p>
          test note 1
        </p>
        <p>
          21/05/21
        </p>
      </div>
    </a>
    <a
      href="/note/6098bbdc04c0fa63070bfe5e"
    >
      <div
        class="allnotes-note"
      >
        <p>
          untitled
        </p>
        <p>
          31/12/21
        </p>
      </div>
    </a>
  </div>
</div>
`;

exports[`Handles network error correctly  1`] = `
<div>
  <div>
    Error! + 
    An error occurred
  </div>
</div>
`;

exports[`Renders loading state correctly 1`] = `
<div>
  <div
    class="ui active text loader"
  >
    Loading...
  </div>
</div>
`;
